snippets/base/admin/adminmodels.py

```diff
@@ -619,31 +620,50 @@ def get_form(self, request, obj=None, **kwargs):
         form.current_user = request.user
         return form
 
-    def make_published(self, request, queryset):
-        clean_queryset = queryset.exclude(status=models.STATUS_CHOICES['Published'])
+    def _action_status_change(self, action, request, queryset):
+        if action == 'publish':
+            status = models.STATUS_CHOICES['Published']
+            no_action_message = 'Skipped {} already published snippets.'
+            success_message = 'Published {} snippets.'
+        else:
+            status = models.STATUS_CHOICES['Draft']
+            no_action_message = 'Skipped {} already unpublished snippets.'
+            success_message = 'Unpublished {} snippets.'
+
+        clean_queryset = queryset.exclude(status=status)
         no_snippets = clean_queryset.count()
         no_already_published_snippets = queryset.count() - no_snippets
 
-        snippets = []
-        with transaction.atomic():
-            for snippet in clean_queryset:
-                snippet.status = models.STATUS_CHOICES['Published']
-                snippet.save()
-                snippets.append(snippet)
+        now = datetime.utcnow()
+        # Create a list of matching snippets before altering them to log action
+        # and send Slack messages.
+        snippets = [snippet for snippet in clean_queryset]
+        clean_queryset.update(status=status, modified=now)
 
         for snippet in snippets:
-            slack.send_slack('asr_published', snippet)
+            self.log_change(request, snippet, 'Changed status.')
```

Comment:
@pmac i remember us talking about logging user actions: all Django Admin actions are logged by default and one can use `self.log_{change,addition,deletion}` to log custom actions, like in this case.

Thought you might find this interesting. Other than that, this PR is mostly moving code around to add an additional snippet unpublish action, next to the existing snippet publish action.